src/lib.rs

```diff
@@ -274,8 +268,12 @@ impl<T: PartialEq> PartialEq for Cactus<T> {
             si.take().map(|n| si = n.parent.as_ref());
             oi.take().map(|n| oi = n.parent.as_ref());
         }
-        debug_assert!(oi.is_none());
-        true
+        if si.is_some() || oi.is_some() {
```

Comment:
Might be nice to expand on the existing comment here.  "we know at least one of {`si`, `oi`} to be `None` due to the loop condition above, now if one of the cacti still has elements, then they have different lengths, thus cannot be equal"

Reply:
Agreed, this could be clearer.